internal/agent/device/hook/manager.go

```diff
@@ -2,270 +2,169 @@ package hook
+	actionCtx := newActionContext(api.DeviceLifecycleHookAfterRebooting, nil, nil, true)
+	return m.loadAndExecuteActions(ctx, actionCtx)
 }
 
-func (m *manager) Sync(currentPtr, desiredPtr *v1alpha1.RenderedDeviceSpec) error {
-	m.log.Debug("Syncing hook manager")
-	defer m.log.Debug("Finished syncing hook manager")
+func (m *manager) loadAndExecuteActions(ctx context.Context, actionCtx *actionContext) error {
+	m.log.Debugf("Starting hook manager On%s()", actionCtx.hook)
+	defer m.log.Debugf("Finished hook manager On%s()", actionCtx.hook)
 
-	current := util.FromPtr(currentPtr)
-	desired := util.FromPtr(desiredPtr)
-	if m.initialized.Load() && reflect.DeepEqual(current.Hooks, desired.Hooks) {
-		m.log.Debug("Hooks are equal. Nothing to update")
-		return nil
-	}
-	desiredHooks := util.FromPtr(desired.Hooks)
-	beforeCreateMap, beforeUpdateMap, beforeRemoveMap, beforeRebootMap, err := m.generateOperationMaps(util.FromPtr(desiredHooks.BeforeUpdating), defaultBeforeUpdateHooks()...)
+	actions, err := m.loadAndMergeActions(actionCtx.hook)
 	if err != nil {
 		return err
 	}
-	afterCreateMap, afterUpdateMap, afterRemoveMap, afterRebootMap, err := m.generateOperationMaps(util.FromPtr(desiredHooks.AfterUpdating), defaultAfterUpdateHooks()...)
-	if err != nil {
-		return err
-	}
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.onBeforeCreate = beforeCreateMap
-	m.onBeforeUpdate = beforeUpdateMap
-	m.onBeforeRemove = beforeRemoveMap
-	m.onAfterCreate = afterCreateMap
-	m.onAfterUpdate = afterUpdateMap
-	m.onAfterRemove = afterRemoveMap
-	m.onBeforeReboot = beforeRebootMap
-	m.onAfterReboot = afterRebootMap
-	m.initialized.Store(true)
-	return nil
+	return m.executeActions(ctx, actions, actionCtx)
 }
 
-func (m *manager) setError(path string, err error) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+func (m *manager) loadAndMergeActions(hookType api.DeviceLifecycleHookType) ([]api.HookAction, error) {
+	actionsMap := map[string][]api.HookAction{}
+	// Read actions from the read-only hooks directory (/usr/lib/flightctl/hooks.d/${hookType}/*.yaml)
+	err := m.loadActions(actionsMap, filepath.Join(ReadOnlyConfigDir, HooksDropInDirName, strings.ToLower(string(hookType)), "*.yaml"))
 	if err != nil {
-		m.errors[path] = err
-	} else {
-		delete(m.errors, path)
+		return []api.HookAction{}, err
 	}
-}
-
-func (m *manager) getActionsForPath(path string, actions ActionMap) []ActionHook {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	return append(actions[path], actions[filepath.Dir(path)]...)
-}
-
-func (m *manager) submitBackgroundJob(path string, actions ActionMap) {
-	actionList := m.getActionsForPath(path, actions)
-	if len(actionList) == 0 {
-		return
+	// Overlay actions from the user-writable hooks directory (/etc/flightctl/hooks.d/${hookType}/*.yaml)
+	err = m.loadActions(actionsMap, filepath.Join(UserWritableConfigDir, HooksDropInDirName, strings.ToLower(string(hookType)), "*.yaml"))
+	if err != nil {
+		return []api.HookAction{}, err
```

Comment:
in a non nil error the object should never be consumed.
```suggestion
		return nil, err
```